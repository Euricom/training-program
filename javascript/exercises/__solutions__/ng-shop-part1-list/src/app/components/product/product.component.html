<div class="panel panel-default">
  <div class="panel-heading">{{product.title}}</div>
  <div class="panel-body">
    <img [src]="product.image" width="100%">
    <table class="table">
      <tbody>
        <tr>
          <td><strong>Id:</strong></td>
          <td>{{product.id}}</td>
        </tr>
        <tr>
          <td><strong>Sku:</strong></td>
          <td>{{product.sku}}</td>
        </tr>
        <tr>
          <td><strong>Title:</strong></td>
          <td>{{product.title}}</td>
        </tr>
        <tr>
          <td><strong>Price:</strong></td>
          <td>{{product.price | currency:'EUR':true}}</td>
        </tr>
        <tr>
          <td><strong>Base Price:</strong></td>
          <td>{{product.basePrice | currency:'EUR':true}}</td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
