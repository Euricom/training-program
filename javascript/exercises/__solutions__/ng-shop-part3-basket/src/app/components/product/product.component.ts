import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { FormControl } from '@angular/forms';

@Component({
  selector: 'product',
  templateUrl: './product.component.html',
  styleUrls: ['./product.component.css'],
})
export class ProductComponent implements OnInit {
  @Input() product: any;
  @Output() addProduct = new EventEmitter();
  quantityControl: FormControl = new FormControl();

  ngOnInit(): void {
    this.quantityControl = new FormControl(1);
  }

  onAdd() {
    const quantity = this.quantityControl.value;
    this.addProduct.emit({
      product: this.product,
      quantity: +quantity,
    });
  }
}
