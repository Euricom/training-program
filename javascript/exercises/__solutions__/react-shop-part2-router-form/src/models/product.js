// eslint-disable-next-line import/prefer-default-export
export class Product {
  constructor(data) {
    Object.assign(this, data);
  }

  get isNew() {
    return !!this.id;
  }

  updateFrom(values) {
    this.sku = values.sku;
    this.price = +values.price;
    this.basePrice = +values.basePrice;
    this.title = values.title;
    this.desc = values.desc;
  }
}
