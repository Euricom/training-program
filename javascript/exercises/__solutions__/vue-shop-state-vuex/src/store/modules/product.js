/* eslint-disable no-param-reassign */
import { GET_PRODUCTS, GET_PRODUCT, ADD_PRODUCT, DELETE_PRODUCT } from '../actionTypes';
import {
  GET_PRODUCTS_SUCCESS,
  GET_PRODUCTS_ERROR,
  GET_PRODUCT_ERROR,
  GET_PRODUCT_SUCCESS,
  ADD_PRODUCT_ERROR,
  ADD_PRODUCT_SUCCESS,
  DELETE_PRODUCT_ERROR,
  DELETE_PRODUCT_SUCCESS,
} from '../mutationTypes';
import productService from '../../services/productService';

const actions = {
  async [GET_PRODUCTS]({ commit, state }, sortParams) {
    try {
      if (!sortParams) {
        sortParams = {};
      }
      if (state.items.length !== 20) {
        console.log('getting all prods');
        const products = await productService.getAll(sortParams.page, sortParams.sortExpression);
        return commit(GET_PRODUCTS_SUCCESS, products);
      }
      return commit(GET_PRODUCTS_SUCCESS);
    } catch (error) {
      return commit(GET_PRODUCTS_ERROR, error.message);
    }
  },
  async [GET_PRODUCT]({ commit, state }, id) {
    try {
      const foundProduct = state.items.find(item => item.id === id);
      if (!foundProduct) {
        const product = await productService.getById(id);
        return commit(GET_PRODUCT_SUCCESS, product);
      }
      return commit(GET_PRODUCTS_SUCCESS);
    } catch (error) {
      return commit(GET_PRODUCT_ERROR, error.message);
    }
  },
  async [DELETE_PRODUCT]({ commit }, id) {
    try {
      const product = await productService.delete(id);
      commit(DELETE_PRODUCT_SUCCESS, product);
    } catch (error) {
      commit(DELETE_PRODUCT_ERROR, error.message);
    }
  },
  async [ADD_PRODUCT]({ commit }, productDto) {
    try {
      if (productDto.id) {
        const product = await productService.update(productDto);
        return commit(ADD_PRODUCT_SUCCESS, product);
      }
      const product = await productService.create(productDto);
      return commit(ADD_PRODUCT_SUCCESS, product);
    } catch (error) {
      return commit(ADD_PRODUCT_ERROR, error.message);
    }
  },
};

const mutations = {
  [GET_PRODUCT_SUCCESS](state, payload) {
    if (payload) return state.items.push(payload);
    return state.items;
  },
  [GET_PRODUCTS_SUCCESS](state, payload) {
    if (payload) {
      // eslint-disable-next-line
      return (state.items = payload);
    }
    return state.items;
  },
  [GET_PRODUCTS_ERROR](state, error) {
    state.error = error;
  },
  [GET_PRODUCT_ERROR](state, error) {
    state.error = error;
  },
  [DELETE_PRODUCT_SUCCESS](state, payload) {
    state.items = state.items.filter(item => item.id !== payload.id);
  },
  [DELETE_PRODUCT_ERROR](state, error) {
    state.error = error;
  },
  [ADD_PRODUCT_SUCCESS](state, payload) {
    const savedItem = state.items.find(item => item.id === payload.id);
    if (savedItem) {
      return state.items.splice(state.items.indexOf(savedItem), 1, payload);
    }
    return state.items.push(payload);
  },
  [ADD_PRODUCT_ERROR](state, error) {
    state.error = error;
  },
};

const getters = {
  products: state => state.items,
  product: state => id => state.items.find(product => product.id === id),
  productsError: state => state.error,
};

const state = () => ({
  error: null,
  items: [],
});

// eslint-disable-next-line
export default {
  actions,
  mutations,
  getters,
  state,
};
