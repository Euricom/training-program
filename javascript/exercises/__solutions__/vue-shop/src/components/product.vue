<template>
  <div>
    <div class="panel panel-default">
      <div class="panel-heading">{{ product.title }}</div>
      <div class="panel-body">
        <img :src="product.image" width="100%" />
        <table class="table">
          <tbody>
            <tr>
              <td><strong>Sku:</strong></td>
              <td>{{ product.sku }}</td>
            </tr>
            <tr>
              <td><strong>Title:</strong></td>
              <td>{{ product.title }}</td>
            </tr>
            <tr>
              <td><strong>Price:</strong></td>
              <td>{{ product.price | currency }}</td>
            </tr>
            <tr>
              <td><strong>Base Price:</strong></td>
              <td>{{ product.basePrice | currency }}</td>
            </tr>
          </tbody>
        </table>
        <button class="btn btn-primary margin" @click="onAdd();" :disabled="!hasProductInStock">Add</button>
        <input v-if="product.stocked" type="number" value="1" style="width: 30px;" min="1" max="5" v-model="quantity" />
        <span v-if="!product.stocked" style="color: red;">Out of Stock</span>
      </div>
    </div>
  </div>
</template>

<script>
import { eventBus } from '@/main';

export default {
  props: ['product'],
  data() {
    return {
      quantity: 1,
    };
  },
  methods: {
    onAdd() {
      eventBus.$emit('addToBasket', { product: this.product, quantity: +this.quantity });
    },
  },
  computed: {
    hasProductInStock() {
      return this.product.stocked;
    },
  },
};
</script>

<style scoped>
.margin {
  margin: 5px;
}
</style>
