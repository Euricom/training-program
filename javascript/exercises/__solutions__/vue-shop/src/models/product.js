/* eslint-disable import/prefer-default-export */

// TODO: add unit test

export class Product {
  constructor(data) {
    if (data) {
      const newProduct = {
        id: data.id,
        sku: data.sku,
        title: data.title,
        price: Number(data.price), // TODO: why number convert?
        basePrice: Number(data.basePrice), // TODO: why number convert?
        desc: data.desc,
        stocked: data.stocked,
        image: data.image,
      };
      return Object.assign(this, newProduct);
    }
    Object.assign(this, {}); // TODO: why is this needed
  }

  isNew() {
    return !this.id;
  }
}
