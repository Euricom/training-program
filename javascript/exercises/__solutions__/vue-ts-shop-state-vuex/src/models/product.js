// eslint-disable-next-line
export class Product {
  // id: number
  // sku: string
  // title: string

  constructor(data) {
    Object.assign(this, data);
  }
}
