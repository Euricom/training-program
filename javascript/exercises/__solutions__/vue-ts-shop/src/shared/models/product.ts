import { IProductDTO } from '../services/productService';

export class Product {
  id?: number;
  sku?: string;
  title?: string;
  desc?: string;
  stocked?: boolean;
  basePrice?: number;
  price?: number;
  image = '/assets/defaultProduct.png';

  constructor(data?: IProductDTO) {
    if (data) {
      // copy over all properties
      Object.assign(this, data);
    }
  }

  isNew() {
    return !this.id;
  }
}
