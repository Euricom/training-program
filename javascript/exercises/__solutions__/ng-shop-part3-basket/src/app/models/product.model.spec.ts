import { Product } from './product.model';

describe('Product', () => {
  test('DTO mapping', () => {
    const prod = new Product({
      id: 123,
      sku: 'abc',
      title: 'aaaaaa',
      desc: 'bbbbb',
      stocked: false,
      basePrice: 0,
      price: 0,
    });
    expect(prod).toMatchSnapshot();
  });

  test('Empty product', () => {
    const prod = new Product();
    expect(prod.id).toBe(0);
    expect(prod.image).toEqual('/assets/defaultProduct.png');
  });
});
