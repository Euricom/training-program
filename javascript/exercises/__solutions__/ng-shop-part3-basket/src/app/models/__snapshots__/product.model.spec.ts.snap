// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product DTO mapping 1`] = `
Product {
  "basePrice": 0,
  "desc": "bbbbb",
  "id": 123,
  "image": "/assets/defaultProduct.png",
  "price": 0,
  "sku": "abc",
  "stocked": false,
  "title": "aaaaaa",
}
`;
